### ISSUE TITLE
This extension works on Firefox

### ISSUE THREAD

**Issue by haxibami** [dimdenGD/OldTweetDeck#9]
# Procedure (after installation)

Note: I tested this with `OldTweetDeckFirefox.zip` **v1**, generated by `npm run build`.

Note2: This procedure is for **v1**, not v2

## 1. Grant permissions manually

In the current implementation of Manifest V3, Firefox [treats](https://discourse.mozilla.org/t/request-permissions-at-install-time-manifest-v3/121543/2) some permissions listed in `host_permissions` as "optional". So users need to **manually** grant them from "`about:addons` > Extensions > Old TweetDeck > Manage > Permissions".

![manually granting permissions](https://github.com/dimdenGD/OldTweetDeck/assets/57034105/f209935d-d920-42e3-abee-a3991db75757)

## 2. Disable tracking protection

Also, Firefox's "Enhanced Tracking Protection" prevents `<iframe>` elements (inserted by this extension) from loading correctly. You can add exception for TweetDeck.

![disabling tracking protection](https://github.com/dimdenGD/OldTweetDeck/assets/57034105/c79c645e-62d3-4e1f-9822-a0ccd5fcb38e)

**Comment by archeruwu** [comment]
How were you able to install it?

**Comment by haxibami** [comment]
The author provides a build script (`pack.js`) for Firefox. You can run it to generate a patched `.zip` file.

```sh
git clone https://github.com/dimdenGD/OldTweetDeck.git
cd OldTweetDeck
npm install
npm run build
mv ../OldTweetDeckFirefox.zip ./
```

You should then find `OldTweetDeckFirefox.zip`. Try temporary installation (unsigned) or self-distribution (signed), as you wish.

**Comment by archeruwu** [comment]
Thank you!

**Comment by Koraydi** [comment]
> 

When I try to run the pack.js file I get this error

![image](https://github.com/dimdenGD/OldTweetDeck/assets/91707942/688029cc-5af4-4f5a-9568-eb23d2084b2d)

**Comment by archeruwu** [comment]
### Heads up!
**You will need Firefox _Developer_ or _Nightly_ to add unsigned extensions.**

**Comment by haxibami** [comment]
> > 
> 
> 
> 
> When I try to run the pack.js file I get this error
> 
> 
> 
> ![image](https://github.com/dimdenGD/OldTweetDeck/assets/91707942/688029cc-5af4-4f5a-9568-eb23d2084b2d)
> 
> 
> 
> 

The `pack.js` file seems to be running as JScript, not JavaScript. How did you run it? do you have Node.js installed?

**Comment by Koraydi** [comment]
> > > 
> > 
> > 
> > When I try to run the pack.js file I get this error
> > ![image](https://user-images.githubusercontent.com/91707942/261784484-688029cc-5af4-4f5a-9568-eb23d2084b2d.png)
> 
> The `pack.js` file seems to be running as JScript, not JavaScript. How did you run it? do you have Node.js installed?

I double clicked on the file in the folder and it gave that error. I have installed Node.js and attempted to run the script with it but it opens for a split second and then closes after doing seemingly nothing.

**Comment by dimdenGD** [comment]
check parent folder of folder it's in, archives probably appeared in it

**Comment by Koraydi** [comment]
![image](https://github.com/dimdenGD/OldTweetDeck/assets/91707942/77e0ca7c-d18e-4b61-9b2d-6405cbdfe35b)
I managed to catch it with a screen recorder

**Comment by dimdenGD** [comment]
run `npm install` in the folder

**Comment by dimdenGD** [comment]
I've built version and made a Release, check https://github.com/dimdenGD/OldTweetDeck/releases/tag/v1.0.2

**Comment by natsuki72855** [comment]
only for developer/Nightly edition? 
Can I use unsigned add-ons with the regular version?

**Comment by Lisandra-dev** [comment]
Sorry, I feel stupid, but I can't understand how to "add" the zip into firefox :s. When I try to load it (using developer version or my normal version) I get an error.
I downloaded the zip version in the BetterTweetdeck fork page.

**Comment by PRGilland** [comment]
When I install it on Firefox Nightly, I get this when I reload the tab, even though I'm still logged in when I get redirected:
![firefox_2023-08-19_14-40-46](https://github.com/dimdenGD/OldTweetDeck/assets/53190048/d68a8ce6-2175-4433-87ae-42226337e65f)

**Comment by Lisandra-dev** [comment]
I need to refresh a lot to get it works, on my side!

**Comment by archeruwu** [comment]
If you get the login error, or it just doesn't load, refresh a few times.

**Comment by PRGilland** [comment]
I don't know how many "a few" is, but it's still not working. If this is something that has to be done every new session, It's probably not worth it for the time being

**Comment by archeruwu** [comment]
Any way of getting legacy back is still better than dealing with "XPro" to be honest.

**Comment by LisandraBrave** [comment]
`manifest.json` lacks an add-on ID in `"browser_specific_settings"`, meaning it will only install through `about:debugging` on firefox (even on nightly and developer), and won't install permanently.
However, reportedly chrome doesn't support the `"browser_specific_settings"` key and will show a warning if it's included. One possible solution is separate `manifest.json` per platform

**Comment by dimdenGD** [comment]
This is already done in `pack.js`

**Comment by dimdenGD** [comment]
You can only install unsigned extensions on Firefox Beta/Nightly

**Comment by LisandraBrave** [comment]
Does it not work on developer (which lists being on aurora update channel)?

**Comment by archeruwu** [comment]
Anyone else getting near unusable lag after leaving the tab open for an extended amount of time? (Firefox Dev Edition)

**Comment by Fapo39** [comment]
still doesn't work for Firefox Nightly or ERS even with v2.02 while it works just fine for Chrome

**Comment by haxibami** [comment]
In my case, v2.0.2 does not work on Firefox. `<script>` elements injected by `element.innerHTML` in `injection.js` are blocked by content security policy (CSP).

![image](https://github.com/dimdenGD/OldTweetDeck/assets/57034105/ee6b0d8c-df09-46e8-9320-c6160844b8c3)

So I tried loading these scripts (`bundle.js`, `vendor.js`) as external files (`element.src`) - this time, they load fine, but the use of `eval()` in them violates CSP. 

In any case, `Content-Security-Policy` header does not seem to be removed, despite the existence of the [relevant code](https://github.com/dimdenGD/OldTweetDeck/blob/13f35b2cf128c7e1ca972296f609348f35951e3c/background.js#L1-L10).

![image](https://github.com/dimdenGD/OldTweetDeck/assets/57034105/b104dfe4-0643-4f74-a5fc-164570c0beb7)

**Comment by haxibami** [comment]
OK. Found the cause:

In Firefox, `extraInfoSpec` in `browser.webRequest.onHeadersReceived.addListener` only [allows](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/webRequest/onHeadersReceived#addlistener_syntax) the values `"blocking"` or `"responseHeaders"` (or both) ,while in Chrome `"extraHeaders"` is also [supported](https://developer.chrome.com/docs/extensions/reference/webRequest/#type-OnHeadersReceivedOptions).

So we need some switching like this:

```diff
diff --git a/background.js b/background.js
index 235f726..d9ddd78 100644
--- a/background.js
+++ b/background.js
@@ -1,3 +1,11 @@
+const extraInfoSpec = ["blocking", "responseHeaders"];
+
+if (
+    chrome.webRequest.OnHeadersReceivedOptions.hasOwnProperty("EXTRA_HEADERS")
+) {
+    extraInfoSpec.push("extraHeaders");
+}
+
 chrome.webRequest.onHeadersReceived.addListener(
     function(details) {
         let headers = details.responseHeaders.filter(header => header.name.toLowerCase() !== 'content-security-policy' && header.name.toLowerCase() !== 'location');
@@ -6,7 +14,7 @@ chrome.webRequest.onHeadersReceived.addListener(
         }
     },
     {urls: ["https://tweetdeck.twitter.com/*"]},
-    ["blocking", "responseHeaders", "extraHeaders"]
+    extraInfoSpec
 );
 
 chrome.webRequest.onBeforeRequest.addListener(
```

**Comment by dimdenGD** [comment]
Thanks, I've updated repo and released new version

**Comment by Dajova** [comment]
Sorry for bringing this up, but it either seems broken again or it was just temporarily.

![image](https://i.imgur.com/eEgdfej.png)

All i get is this, even after clearing cookies, reinstalling the extension several times and refreshing the page once a minute for almost a hour. It worked the first time, but not once since then.

**Comment by Audrican** [comment]
> You can only install unsigned extensions on Firefox Beta/Nightly

You can install it on main firefox as a "temporary extension" in `about:debugging` > This Firefox.
It'll uninstall upon closing but it's still an alternative for main fox.
